src/compiler/rust.rs

```diff
@@ -0,0 +1,761 @@
+        if output.status.success() {
+            let outstr = String::from_utf8(output.stdout).chain_err(|| "Error parsing rustc output")?;
+            Ok(outstr.lines().map(|l| l.to_owned()).collect())
+        } else {
+            bail!("Failed to run `rustc --print file-names`");
+        }
+    }))
+}
+
+impl<T> Compiler<T> for Rust
+    where T: CommandCreatorSync,
+{
+    fn kind(&self) -> CompilerKind { CompilerKind::Rust }
+    /// Parse `arguments` as rustc command-line arguments, determine if
+    /// we can cache the result of compilation. This is only intended to
+    /// cover a subset of rustc invocations, primarily focused on those
+    /// that will occur when cargo invokes rustc.
+    ///
+    /// Caveats:
+    /// * We don't support compilation from stdin.
+    /// * We require --emit.
+    /// * We only support `link` and `dep-info` in --emit (and don't support *just* 'dep-info')
+    /// * We require `--out-dir`.
+    /// * We don't support `-o file`.
+    fn parse_arguments(&self,
+                       arguments: &[String],
+                       cwd: &Path) -> CompilerArguments<Box<CompilerHasher<T> + 'static>>
+    {
+        match parse_arguments(arguments, cwd) {
+            CompilerArguments::Ok(h) => CompilerArguments::Ok(Box::new(h)),
+            CompilerArguments::NotCompilation => CompilerArguments::NotCompilation,
+            CompilerArguments::CannotCache => CompilerArguments::CannotCache,
+        }
+    }
+
+
+    fn box_clone(&self) -> Box<Compiler<T>> {
+        Box::new((*self).clone())
+    }
+}
+
+fn parse_arguments(arguments: &[String], _cwd: &Path) -> CompilerArguments<RustHasher>
+{
+    //TODO: use lazy_static for this.
+    let args_with_val: HashSet<&'static str> = HashSet::from_iter(ARGS_WITH_VALUE.iter().map(|v| *v));
+    let mut emit: Option<HashSet<&str>> = None;
+    let mut input = None;
+    let mut output_dir = None;
+    let mut crate_name = None;
+    let mut extra_filename = None;
+    let mut externs = vec![];
+    let mut filtered_arguments = vec![];
+
+    let mut it = arguments.iter();
+    while let Some(arg) = it.next() {
+        match arg.as_str() {
+            // Various non-compilation options.
+            "--help" | "-V" | "--version" => return CompilerArguments::NotCompilation,
+            v if v.starts_with("--print") || v.starts_with("--explain") || v.starts_with("--pretty") || v.starts_with("--unpretty") => return CompilerArguments::NotCompilation,
+            // Could support `-o file` but it'd be more complicated.
+            "-o" => return CompilerArguments::CannotCache,
+            //TODO: support linking against native libraries. This
+            // will require replicating the linker search strategy
+            // so we can *find* them.
+            "-l" => return CompilerArguments::CannotCache,
+            v if v.starts_with("--emit") => {
+                //XXX: do we need to handle --emit specified more than once?
```

Comment:
Yeah I think it's fine to just return "not cacheable" for multiple `--emit` calls for now.